viewing_party/party.py

```diff
@@ -1,23 +1,168 @@
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
 
-# -----------------------------------------
-# ------------- WAVE 3 --------------------
-# -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum = 0
+    ratings_list = []
+    if user_data['watched']:
+        for movie in user_data["watched"]:
+            ratings_list.append(movie['rating'])
+        for movie in ratings_list:
+            sum += movie
+        average = sum/len(ratings_list)
+        return average
+    return 0.0
 
-        
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
 
-# -----------------------------------------
-# ------------- WAVE 5 --------------------
-# -----------------------------------------
+def get_most_watched_genre(user_data):
+    genre_list = []
+    genre_tally_dict = {}
+    most_watched_str = None
+    most_watched_int = 0
+    if user_data:
+        for i in user_data["watched"]:
+            genre_list.append(i['genre'])
+        for i in genre_list:
+            if i not in genre_tally_dict:
+                genre_tally_dict[i] = 1
+                most_watched_int = 1
+            else:
+                genre_tally_dict[i] += 1
+                most_watched_int += 1
+            for i in genre_tally_dict:
+                if genre_tally_dict[i] > most_watched_int:
+                    most_watched_str = i
+        return most_watched_str
+    return None
+
+    # -----------------------------------------
+    # ------------- WAVE 3 --------------------
+    # -----------------------------------------
+
+
+def get_unique_watched(user_data):
+    user_watched_list = user_data['watched']
+    friends = user_data['friends']
+    user_unique_watched_list = []
+    friends_title_list = []
+    for friend in friends:
+        for movie in friend['watched']:
+            friends_title_list.append(movie['title'])
+    for movie in user_watched_list:
+        if movie['title'] not in friends_title_list:
+            user_unique_watched_list.append(movie)
+    return user_unique_watched_list
+
+
+def get_friends_unique_watched(user_data):
+
+    user_watched_list = user_data['watched']
+    friends = user_data['friends']
+    friends_unique_movie_list = []
+    user_title_list = []
+    for movie in user_watched_list:
+        user_title_list.append(movie['title'])
+    for friend in friends:
+        for movie in friend['watched']:
+            if movie['title'] not in user_title_list and movie not in friends_unique_movie_list:
+                friends_unique_movie_list.append(movie)
+    return friends_unique_movie_list
+
+    # -----------------------------------------
+    # ------------- WAVE 4 --------------------
+    # -----------------------------------------
+
+
+def get_available_recs(user_data):
+
+    subscriptions_list = user_data['subscriptions']

```

Comment:
If we built a `set` from the subscription list, the host check later on would be potentially much more efficent (using set membership checks rather than needing to iterate through a list each time)

```py
    subscriptions = set(user_data['subscriptions'])
```